aten/src/ATen/native/cudnn/RNN.cpp

```diff
@@ -0,0 +1,896 @@
+      if (layer == 0) {
+        global_layer_params_count = layer_params_count;
+      } else {
+        AT_ASSERT(global_layer_params_count == layer_params_count, "%d (global) != %d", global_layer_params_count, layer_params_count);
+      }
+    } // for layer
+    return std::make_pair(params, global_layer_params_count);
+  }
+
+  void _copyParams(MatrixRef<Tensor> params_from, MatrixRef<Tensor> params_to) {
+    AT_ASSERT(params_from.size(0) == params_to.size(0), "number of layers mismatch");
+    for (size_t i = 0; i < params_from.size(0); i++) {
+      auto layer_params_from = params_from[i];
+      auto layer_params_to = params_to[i];
+      for (auto a = layer_params_from.begin(), b = layer_params_to.begin();
+           a != layer_params_from.end() && b != layer_params_to.end();
+           ++a, ++b) {
+        auto param_from = *a, param_to = *b;
+        AT_ASSERT(param_from.type() == param_to.type(), "parameter types mismatch");
+        param_to.copy_(param_from.view_as(param_to));
+      }
+    }
+  }
+
+  std::vector<int64_t> _input_size(const TensorDescriptorListParams& tensors) {
+    if (tensors.is_input_packed()) {
+      return {tensors.outer_size, tensors.inner_size};
+    } else {
+      return {tensors.seq_length, tensors.mini_batch, tensors.inner_size};
+    }
+  }
+
+  std::vector<int64_t> _hidden_size(const RNNDescriptorParams& rnn, const TensorDescriptorListParams& tensors) {
+    return {rnn.num_layers * rnn.num_directions(), tensors.mini_batch, rnn.hidden_size};
+  }
+
+  std::vector<int64_t> _output_size(const RNNDescriptorParams& rnn, const TensorDescriptorListParams& tensors) {
+    if (tensors.is_input_packed()) {
+      return {tensors.outer_size, rnn.hidden_size * rnn.num_directions()};
+    } else {
+      return {tensors.seq_length, tensors.mini_batch, rnn.hidden_size * rnn.num_directions()};
+    }
+  }
+
+} // anonymous namespace
+
+// NB: does inplace update into TensorList
+// It would be a relatively simple matter to refactor this into multiple
+// functions, only one of which does an inplace update, but we leave this
+// for future work
+Tensor _cudnn_rnn_flatten_weight(
+    TensorList weight_arr, int64_t weight_stride0,
+    int64_t input_size,
+    int64_t fn_mode, int64_t fn_hidden_size,
+    int64_t fn_num_layers, bool batch_first,
+    bool fn_bidirectional
+    ) {
+
+  if (weight_arr.size() == 0) {
+    throw std::runtime_error("_cudnn_rnn_flatten_weight_: cannot flatten empty weight list");
+  }
+
+  auto any_param = weight_arr[0];
+
+  RNNDescriptorParams rnn;
+  rnn.set_mode(fn_mode);
+  rnn.hidden_size = fn_hidden_size;
+  rnn.num_layers = fn_num_layers;
+  rnn.set_bidirectional(fn_bidirectional);
+  rnn.datatype = getCudnnDataType(any_param);
+
+  auto handle = getCudnnHandle();
+  // NB: So, I am pretty sure that get_parameters() does not rely in any way
+  // on the dropout descriptor.  So we fake up a dummy one instead of try
+  // to actually make one legitimately.
```

Comment:
but why?

Reply:
It's mostly to keep `cudnn_rnn_flatten_weight`s API simpler, by letting it avoid passing in the dropout descriptor state (which wouldn't actually do anything.)

I'll move this comment to a better call site.